opentelemetry-sdk/tests/trace/test_span_processor.py

```diff
@@ -0,0 +1,314 @@
+
+        flushed = multi_processor.force_flush(50)
+        self.assertFalse(flushed)
+        self.assertEqual(1, mock_processor1.force_flush.call_count)
+        self.assertEqual(0, mock_processor2.force_flush.call_count)
+
+
+class TestConcurrentMultiSpanProcessor(unittest.TestCase):
+    def test_on_start(self):
+        multi_processor = trace.ConcurrentMultiSpanProcessor(3)
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        multi_processor.on_start(mock.Mock(spec=trace.Span))
+
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.on_start.call_count)
+        multi_processor.shutdown()
+
+    def test_on_end(self):
+        multi_processor = trace.ConcurrentMultiSpanProcessor(3)
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+        mock_span = mock.Mock(spec=trace.Span)
+
+        multi_processor.on_end(mock_span)
+
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.on_end.call_count)
+        multi_processor.shutdown()
+
+    def test_on_shutdown(self):
+        multi_processor = trace.ConcurrentMultiSpanProcessor(3)
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        multi_processor.shutdown()
+
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.shutdown.call_count)
+
+    def test_force_flush(self):
+        multi_processor = trace.ConcurrentMultiSpanProcessor(3)
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        flushed = multi_processor.force_flush(100)
+
+        self.assertTrue(flushed)
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.force_flush.call_count)
+        multi_processor.shutdown()
+
+    def test_force_flush_late_by_timeout(self):
+        multi_processor = trace.ConcurrentMultiSpanProcessor(5)
+        wait_event = Event()
+
+        def delayed_flush(_):
+            wait_event.wait()
+
+        late_mock = mock.Mock(spec=trace.SpanProcessor)
+        late_mock.force_flush = mock.Mock(side_effect=delayed_flush)
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 4)]
+        mocks.insert(0, late_mock)
+
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        flushed = multi_processor.force_flush(timeout_millis=10)
+        # let the thread executing the late_mock continue
+        wait_event.set()
+
+        self.assertFalse(flushed)
```

Comment:
this assert could move on line 290?

Reply:
i think this could become troublesome if the assert actually fails because the `ThreadPoolExecutor` is joining all its threads in an `atexit` handler. So this would cause the test process to run indefinitely since noone is setting the `wait_event`.